Documentation/Yarn-Spec.md

```diff
@@ -0,0 +1,636 @@
+For example if `"hello" < 5` is `true` then `5 > "hello"` should also be `true`.
+Likewise `("hello" + 2) < 5` resulted in `true` then `(2 + "hello") < 5` should also be `true`.
+This taken to the extreme should mean that `1 + "hello" == "hello + 1` should evaluate to `true` which is confusing to most people, hence the recommendation against supporting operations between disparate types.
+
+#### Operation Output
+
+The following table shows the expected output type of each operation based on its operand type:
+
+|         | +      | -      | *      | /      | %      | ==      | !=      | gt      | lt      | gte     | lte     | \|\|    | &&      | !       | unary minus |
+|---------|--------|--------|--------|--------|--------|---------|---------|---------|---------|---------|---------|---------|---------|---------|-------------|
+| number  | number | number | number | number | number | boolean | boolean | boolean | boolean | boolean | boolean | boolean | boolean |         | number      |
+| string  | string |        |        |        |        | boolean | boolean |         |         |         |         |         |         |         |             |
+| boolean |        |        |        |        |        | boolean | boolean |         |         |         |         | boolean | boolean | boolean |             |
+
+#### Order of Operations
+
+The order of operations is as follows:
+
+1. `()`
+1. `!`, `unary minus`
+1. `*`, `/`, `%`
+1. `-`, `+`
+1. `==`, `!=`, `lt`, `gt`, `lte`, `gte`
+1. `||`, `&&`
+
+If there are any equal priority operations in an expression they are resolved left to right as encountered in the expression.
+
+### Functions
+
+_Functions_ are an alternate way of getting values into expressions.
+Functions are intended to be used to allow more complex code be bundled and called in a different environment, such as in the game itself.
+Functions must return a value.
+
+#### Structure 
+
+Functions are comprised of a name, parentheses, and parameters.
+
+The _function name_ can be any combination of the characters `a` - `z`, `A` - `Z`, and `0` - `9`, but must not start with a number.
+The minimum and maximum length of function names is unspecified.
+
+The parentheses go after the function name and there must be no whitespace between the opening parentheses `(` and the function name.
+The closing parethensis `)` finishes the function.
+
+_Parameters_ go in between the opening `(` and closing `)` parentheses.
+Parameters must be expressions, functions, values, or variables.
+Functions can have multiple parameters, these must be separated by the comma `,` symbol.
+
+Whitespace between parameters and the separator is undefined. hmm this allows for newlines is that ok?
+The maximum and minimum number of allowed parameters a function can have is undefined.
+
+Examples of functions include the following;
+
+```yarn
+getPlayerName()
+DetermineCurrentRoom($playerName, $target, 2)
+rad2Deg(1.5707963268)
+```
```

Comment:
As per line 454, this example is invalid

Reply:
these are examples of function to be used in expressions, not lines in isolation.

Reply:
This isn't clear from the formatting imo